### ISSUE TITLE
catnum_to_list function for report/label generation

### ISSUE THREAD

**Issue by jtgiermakowski** [ArctosDB/arctos#3937]
Hello,

I'm trying to generate a report for a jar label. I've followed the SQL examples and tweaked the .cfr file to my liking but there is still a persistent issue of sorting the list of specimens by catalog number. 

The following SQL works fine: 

```
select 
  flat.scientific_name, 
  flat.family, 
  flat.country,
  flat.state_prov,
  flat.spec_locality, 
  flat.county, 
  flat.cat_num,
  contains_part_container.label container_label 
FROM 
  flat,
  specimen_part,
  coll_obj_cont_hist,
  container part_container,
  container contains_part_container
WHERE 
  flat.collection_object_id=specimen_part.derived_from_cat_item and
  specimen_part.collection_object_id=coll_obj_cont_hist.collection_object_id and
  coll_obj_cont_hist.container_id=part_container.container_id and
  part_container.parent_container_id=contains_part_container.container_id and
  contains_part_container.container_id in (#container_id#)
ORDER BY
  catalognumberint ASC
```

and generates a table with properly sorted specimens:

<html><body>

  | CAT_NUM | CONTAINER_LABEL | COUNTRY | COUNTY | FAMILY | SCIENTIFIC_NAME | SPEC_LOCALITY | STATE_PROV
-- | -- | -- | -- | -- | -- | -- | -- | --
1 | 10 | MSBHERP:JAR9999 | United States | Lincoln County | Viperidae | Crotalus viridis | 7 miles Northwest Ramon on Highway 285 | New Mexico
2 | 4086 | MSBHERP:JAR9999 | United States | Hidalgo County | Viperidae | Crotalus viridis | 5 miles South and 9 miles East Lordsburg | New Mexico
3 | 13822 | MSBHERP:JAR9999 | United States | Hidalgo County | Viperidae | Crotalus viridis | 11 miles South Animas | New Mexico
4 | 18744 | MSBHERP:JAR9999 | United States | Hidalgo County | Viperidae | Crotalus viridis | North Antelope Wells | New Mexico
5 | 22076 | MSBHERP:JAR9999 | United States | Hidalgo County | Viperidae | Crotalus viridis | 12.5 miles North from Redrock turnoff US 70 | New Mexico
6 | 30525 | MSBHERP:JAR9999 | United States | Hidalgo County | Viperidae | Crotalus viridis | 17 road miles South of Animas on State Road 339 | New Mexico
7 | 32219 | MSBHERP:JAR9999 | United States | Hidalgo County | Viperidae | Crotalus viridis | Animas Valley, Cowan Ranch | New Mexico
8 | 37766 | MSBHERP:JAR9999 | United States | Hidalgo County | Viperidae | Crotalus viridis | 21 kilometers South and 0.3 kilometers West of Animas (center) on NM 338 | New Mexico
9 | 42751 | MSBHERP:JAR9999 | United States | Hidalgo County | Viperidae | Crotalus viridis | NM Highway 464, 0.4 Kilometers North Junction US Highway 70 | New Mexico

</body>
</html>

however, it **does not work** with a "Pre-Function" called _catnum_to_list_  

The resulting output is seemingly sorted alphabetically by locality (!?). This "Pre-function" (at  arctos/Reports/functions/label_functions.cfm ) is very useful because it checks for duplicates on counties, taxonomy, etc. I'd like to keep the pre-function but also have the output ordered by catalognumberint ASC

I'm doing this for **MSB_Herp_Jar_55ISOH** but same problem persists if I try it on another report (like **UCM_JAR_consolidated_state_SMALL**)

Is that the problem with the _catnum_to_list_ function or is there another SQL way to solve the issue?###  thanks!

**Comment by dustymc** [comment]
I don't think there's a straightforward way for the sort to survive the function.

>checks for duplicates

I might be able to help rewrite the SQL if you can expand on that - are you just looking for repeated values in some columns, or merging multiple records into a single label,  or ???

**Comment by jtgiermakowski** [comment]
Yes, ideally I'd want to have the capability to check for different state, county, or species. It's encoded in the .cfr file as follows:

```
IIF((distinct_taxon_count EQ 1),DE(scientific_name),DE("ERROR - mixed taxa"))

country & ": " & IIF((distinct_state_count LTE 1),DE(state_prov),DE("[various regions]")) & ": " & county
```

Of course in the .cfr file, it's looking for the output from the function _catnum_to_list_:

> 
> 			* new column distinct_county_count (count of distinct values in COUNTY)
> 			* new column distinct_state_count (count of distinct values in STATE_PROV)
> 			* new column distinct_taxon_count (count of distinct values in SCIENTIFIC_NAME)

I tried getting access to variable _catnumlist_ (also part of the function _catnum_to_list_ output) but couldn't succeed with SQL....

Thanks!

**Comment by dustymc** [comment]
I'm not sure I quite get what you're doing yet, but you can use string_agg() to aggregate - something like this:

```
select 
  string_agg(flat.scientific_name,'; ') as scientific_name, 
  string_agg(flat.family,'; ') as family, 
  string_agg(flat.country,'; ') as country, 
  string_agg(flat.state_prov,'; ') as state_prov, 
  string_agg(flat.spec_locality,'; ') as spec_locality, 
  string_agg(flat.county,'; ') as county, 
  string_agg(flat.cat_num,'; ') as cat_num, 
  contains_part_container.label container_label 
FROM 
  flat
  inner join specimen_part on flat.collection_object_id=specimen_part.derived_from_cat_item
  inner join coll_obj_cont_hist on specimen_part.collection_object_id=coll_obj_cont_hist.collection_object_id
  inner join container part_container on coll_obj_cont_hist.container_id=part_container.container_id
  inner join container contains_part_container on part_container.parent_container_id=contains_part_container.container_id
WHERE 
  contains_part_container.container_id in (#container_id#)
group by contains_part_container.label 
```

**Comment by dustymc** [comment]
Happy?
